### PR TITLE
feat: Transform strings with respect to property schema when conforming properties

### LINKED ISSUE #3014: feat: Conform primitive properties from strings
### Discussed in https://github.com/meltano/sdk/discussions/2994

<div type='discussions-op-text'>

<sup>Originally posted by **ReubenFrankel** April 22, 2025</sup>
Why does the SDK conform primitive properties of a record dict towards a JSON compatible format (e.g. `datetime.datetime` to `str`), but not from strings to their expected schema type? Is/should it be the responsibility of the developer to define this mapping?

https://github.com/meltano/sdk/blob/c152403ba539a11d0d2b4a699c0885b31d4c7567/singer_sdk/helpers/_typing.py#L521-L547

i.e.

```py
    if isinstance(elem, str) and th.IntegerType.__type_name__ in property_schema["type"]:
        return int(elem)
```</div>

### PR DESCRIPTION
**Pull request by ReubenFrankel**
Adds support for transforming string values to their defined property schema types.

---

Related discussion: https://github.com/meltano/sdk/discussions/2994
Closes https://github.com/meltano/sdk/issues/3014

## Summary by Sourcery

Add automatic type conversion for string properties based on their schema during data conforming.

New Features:
- Transform string values to boolean, integer, number, array, or object types according to the property's JSON schema definition.
- Handle empty strings appropriately for each target type (e.g., 0 for integer/number, False for boolean, empty list/dict for array/object).
- Treat 'inf' and 'nan' strings as null when converting to numbers.
- Parse JSON strings when converting to array or object types.
- Perform case-insensitive matching for 'true' when converting to boolean; other strings become False.
- Return None for empty strings if the target schema is nullable and not string type itself

Tests:
- Add unit tests to cover various string transformation scenarios, including edge cases.

### CODE REVIEW

File: singer_sdk/helpers/_typing.py [hunk 1]
```diff
@@ -542,6 +543,31 @@ def _conform_primitive_property(  # noqa: PLR0911
...
+
+    if is_boolean_type(property_schema):
+        return (
+            elem.lower() == "true"
+        )  # false for any non-"true" string (case-insensitive), including empty string
+    if is_integer_type(property_schema):
+        return int(elem or 0)  # 0 for empty string
+    if is_number_type(property_schema):
+        d = decimal.Decimal(elem or 0)  # 0 for empty string
+        return None if math.isnan(d) or math.isinf(d) else d
```
[the remark this answers is not in the document]
Comment (ReubenFrankel) [reply singer_sdk/helpers/_typing.py]: Actually, we can just use `d.is_finite()` here: https://github.com/meltano/sdk/pull/2997/commits/cc2355958e9c6cbdb381fbdb2f7be21bed9eca09

File: singer_sdk/helpers/_typing.py [hunk 2]
```diff
@@ -542,6 +543,31 @@ def _conform_primitive_property(  # noqa: PLR0911
...
+) -> t.Any:  # noqa: ANN401
+    if not elem and is_null_type(property_schema):
+        return None  # if nullable, None for empty string
+
+    if is_boolean_type(property_schema):
+        return (
+            elem.lower() == "true"
+        )  # false for any non-"true" string (case-insensitive), including empty string
+    if is_integer_type(property_schema):
+        return int(elem or 0)  # 0 for empty string
```
[the remark this answers is not in the document]
Comment (ReubenFrankel) [reply singer_sdk/helpers/_typing.py:563]: There is no existing precedent set for handling exceptions like this yet. Perhaps this can be tackled as part of the [proposed refactor of `_conform_record_data_types`](https://github.com/meltano/sdk/blob/5273f325bd518842290e16673fc33f26c245f1b0/singer_sdk/helpers/_typing.py#L419).

File: tests/core/test_typing.py [hunk 3]
```diff
@@ -353,12 +354,68 @@ def test_conform_object_additional_properties():
...
+            id="string_not_true_to_boolean",
+        ),
+        pytest.param("", {"type": "boolean"}, False, id="string_empty_to_boolean"),
+        pytest.param("3", {"type": "integer"}, 3, id="string_integer_to_integer"),
+        pytest.param("", {"type": "integer"}, 0, id="string_empty_to_integer"),
+        pytest.param(
+            "3.14",
+            {"type": "number"},
+            decimal.Decimal("3.14"),
+            id="string_float_to_number",
```
[the remark this answers is not in the document]
Comment (ReubenFrankel) [reply tests/core/test_typing.py:385]: See https://github.com/meltano/sdk/pull/2997#discussion_r2060391147

File: tests/core/test_typing.py [hunk 4]
```diff
@@ -353,12 +354,68 @@ def test_conform_object_additional_properties():
...
+            "",
+            {"type": "number"},
+            decimal.Decimal(0),
+            id="string_empty_to_number",
+        ),
+        pytest.param(
+            "[1, 2, 3]",
+            {"type": "array"},
+            [1, 2, 3],
+            id="string_json_array_to_array",
```
[the remark this answers is not in the document]
Comment (ReubenFrankel) [reply tests/core/test_typing.py:399]: See https://github.com/meltano/sdk/pull/2997#discussion_r2060391147

**Review by edgarrmondragon — COMMENTED** [review]
Thanks @ReubenFrankel!

My main concern with this would've been performance, but the benchmark seems unaffected? I could be missing something though.

**Comment by ReubenFrankel** [comment]
> My main concern with this would've been performance, but the benchmark seems unaffected? I could be missing something though.

It looks like it was slightly affected: https://codspeed.io/meltano/sdk/branches/ReubenFrankel%3Afeat%2Fconform-primitive-interpret-string?uri=tests%2Fcore%2Ftest_typing.py%3A%3Atest_bench_conform_record_data_types

Possibly due to this new check, which will check the property schema type for every `str` value: https://github.com/meltano/sdk/blob/6c0f4748df1e5e094fdc8f2cdc60477df9c2d65a/singer_sdk/helpers/_typing.py#L546

File: singer_sdk/helpers/_typing.py [hunk 5]
```diff
@@ -542,9 +543,32 @@ def _conform_primitive_property(  # noqa: PLR0911
         # for BIT value, treat 0 as False and anything else as True
         return elem != b"\x00" if is_boolean_type(property_schema) else elem.hex()
     if isinstance(elem, (float, decimal.Decimal)):
-        if math.isnan(elem) or math.isinf(elem):
-            return None
-        return elem
+        return elem if math.isfinite(elem) else None
+    if isinstance(elem, str) and not is_string_type(property_schema):
```
Comment (edgarrmondragon) [inline singer_sdk/helpers/_typing.py:547]: (commenting here to start a thread)

> It looks like it was slightly affected: https://codspeed.io/meltano/sdk/branches/ReubenFrankel%3Afeat%2Fconform-primitive-interpret-string?uri=tests%2Fcore%2Ftest_typing.py%3A%3Atest_bench_conform_record_data_types
> 
> Possibly due to this new check, which will check the property schema type for every `str` value:
> 
> https://github.com/meltano/sdk/blob/6c0f4748df1e5e094fdc8f2cdc60477df9c2d65a/singer_sdk/helpers/_typing.py#L546

Do you think there's a way we could prevent this regression? Perhaps caching a mapping of property > is_<type>.

Comment (ReubenFrankel) [reply singer_sdk/helpers/_typing.py:547]: Yes, I was thinking about some kind of type caching also. We did something like that [here](https://github.com/Matatika/tap-callminer/blob/e6d59a22d586e16ad01f859dc0043de5b3eef396/tap_callminer/streams.py#L217-L226), or were you imagining it to be per-property?

Comment (edgarrmondragon) [reply singer_sdk/helpers/_typing.py:547]: Yeah, something like or like the [selection mask](https://github.com/meltano/sdk/blob/c585ac6df164c83eb19f5b6b24fa6a1d6a5e6398/singer_sdk/singerlib/catalog.py#L21) we use internally here.

**Comment by ReubenFrankel** [comment]
@edgarrmondragon Doesn't look like caching helped here (we can revert c391924 if necessary)... Not sure what else we can do.
